### PR TITLE
Add Go AzureIdentityCredentialAdapter, readme and test

### PR DESCRIPTION
**Pull request by Luyunmt**
Add Go AzureIdentityCredentialAdapter, readme and test

### CODE REVIEW

**Comment by jongio** [comment]
@joshfree - Who would you like to review this?

**Comment by jongio** [comment]
@catalinaperalta - Could you please review this?

File: go/azidentity/azure_identity_credential_adapter_test.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter_test.go:1]: Missing MIT license header.

File: go/azidentity/azure_identity_token_provider.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_token_provider.go:1]: Missing MIT license header.

File: go/azidentity/azure_identity_token_provider.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_token_provider.go:31]: The use of panic will preclude just about anybody from adopting this.  It will need to be rewritten to obtain the token earlier where an error can be returned.

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter.go:14]: I don't think the base adapter should default to any kind of credential.  We can build higher-level constructs on top of this that do.

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter.go:12]: 'Aapter' => 'Adapter'

File: go/azidentity/azure_identity_token_provider.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_token_provider.go:21]: Similar to earlier comment, don't default here.

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jongio) [reply go/azidentity/azure_identity_credential_adapter.go:14]: @jhendrixMSFT - Just FYI that all the other langs default to DAC.  Given that, do you still feel like Go shouldn't default to DAC?

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter.go:16]: This needs to adapt an `azcore.Credential` to an `autorest.Authorizer`, then other constructs can build on top of them.  I think it should look like this.

```go
// AzureIdentityCredentialAdapter is used to adapt an azcore.Credential to an autorest.Authorizer
type AzureIdentityCredentialAdapter interface {
	autorest.Authorizer
}

// NewAzureIdentityCredentialAdapter returns an implementation of AzureIdentityCredentialAdapter.
func NewAzureIdentityCredentialAdapter(credential azcore.Credential, options azcore.AuthenticationPolicyOptions) AzureIdentityCredentialAdapter {
	policy := credential.AuthenticationPolicy(options)
	return &policyAdapter{p: policy}
}

type policyAdapter struct {
	p azcore.Policy
}

// WithAuthorization implements the autorest.Authorizer interface for type policyAdapter.
func (ca *policyAdapter) WithAuthorization() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			r, err := p.Prepare(r)
			if err != nil {
				return r, err
			}
			_, err = ca.p.Do(r.Context(), &azcore.Request{Request: r})
			if errors.Is(err, azcore.ErrNoMorePolicies) {
				err = nil
			}
			return r, err
		})
	}
}
```

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter.go:53]: Updated version based on the rewritten adapter.

```go
// DefaultManagementScope is the default credential scope for Azure Resource Management.
const DefaultManagementScope = "https://management.azure.com//.default"

// DefaultAzureCredentialOptions contains credential and authentication policy options.
type DefaultAzureCredentialOptions struct {
	// DefaultCredential contains configuration options passed to azidentity.NewDefaultAzureCredential().
	// Set this to nil to accept the underlying default behavior.
	DefaultCredential *azidentity.DefaultAzureCredentialOptions

	// AuthenticationPolicy contains configuration options passed to the underlying authentication policy.
	// Setting this to nil will use the DefaultManagementScope when acquiring a token.
	AuthenticationPolicy *azcore.AuthenticationPolicyOptions
}

// NewDefaultAzureCredentialAdapter returns an implementation of AzureIdentityCredentialAdapter.
func NewDefaultAzureCredentialAdapter(options *DefaultAzureCredentialOptions) (AzureIdentityCredentialAdapter, error) {
	if options == nil {
		options = &DefaultAzureCredentialOptions{
			AuthenticationPolicy: &azcore.AuthenticationPolicyOptions{
				Options: azcore.TokenRequestOptions{
					Scopes: []string{DefaultManagementScope},
				},
			},
		}
	}
	chain, err := azidentity.NewDefaultAzureCredential(options.DefaultCredential)
	if err != nil {
		return nil, err
	}
	return NewAzureIdentityCredentialAdapter(chain, *options.AuthenticationPolicy), nil
}
```

File: go/azidentity/azure_identity_token_provider.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_token_provider.go:14]: The types here seem unnecessary?

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [reply go/azidentity/azure_identity_credential_adapter.go:16]: You could probably even omit `AzureIdentityCredentialAdapter` and have `NewAzureIdentityCredentialAdapter()` return an `autorest.Authorizer`

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidentity/azure_identity_credential_adapter.go:4]: I'm also unsure how I feel about naming this package `azidentity`.  Is the assumption that folks will use this in lieu of the "real" `azidentity`?  If they want to use the adapter in tandem with some other credential mechanism in the "real" `azidentity` then one of the package names will have to be aliased; not terrible, but would indicate that we didn't think about this scenario.

Comment (jongio) [reply go/azidentity/azure_identity_credential_adapter.go:4]: Should be azidext

File: go/azidentity/azure_identity_credential_adapter.go
Comment (Luyunmt) [reply go/azidentity/azure_identity_credential_adapter.go]: Agree, I have changed it according this, but if using AuthenticationPolicy(), Error handling goes wrong with wrong azure credential. So I think using gettoken() is better.

File: go/azidentity/azure_identity_credential_adapter.go
Comment (Luyunmt) [reply go/azidentity/azure_identity_credential_adapter.go]: Agree, I have changed it according this, but if using AuthenticationPolicy(), Error handling goes wrong with wrong azure credential. So I think using gettoken() is better.

File: go/azidentity/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [reply go/azidentity/azure_identity_credential_adapter.go]: We _must_ use the underlying authentication policy as it handles properly refreshing the credential when it's near expiration.  Can you please clarify the problem with error handling?

Comment (jhendrixMSFT) [reply go/azidentity/azure_identity_credential_adapter.go]: Presumably you're referring to if the provided credentials are invalid.  In the case of an `azidentity.AuthenticationFailedError` error type we need to wrap it in a type that implements the `adal.TokenRefreshError` interface (see below).  This will prevent the retry policy from retrying on invalid credentials.

```go
// WithAuthorization implements the autorest.Authorizer interface for type policyAdapter.
func (ca *policyAdapter) WithAuthorization() autorest.PrepareDecorator {
	return func(p autorest.Preparer) autorest.Preparer {
		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
			r, err := p.Prepare(r)
			if err != nil {
				return r, err
			}
			_, err = ca.p.Do(r.Context(), &azcore.Request{Request: r})
			if errors.Is(err, azcore.ErrNoMorePolicies) {
				return r, nil
			}
			var afe *azidentity.AuthenticationFailedError
			if errors.As(err, &afe) {
				err = &tokenRefreshError{
					inner: afe,
				}
			}
			return r, err
		})
	}
}

// tokenRefreshError implements the adal.TokenRefreshError interface
type tokenRefreshError struct {
	inner error
}

func (t *tokenRefreshError) Error() string {
	return t.inner.Error()
}

func (t *tokenRefreshError) Response() *http.Response {
	return nil
}

func (t *tokenRefreshError) Unwrap() error {
	return t.inner
}
```

File: README.md
Comment (catalinaperalta) [inline README.md:176]: is the import path staying like this? From the comments I saw the the package will be named `azidext` if so I think it might be more conventional for the path to also end in azidext.

File: go/azure_identity_credential_adapter.go
Comment (catalinaperalta) [inline go/azure_identity_credential_adapter.go:19]: Shouldn't the user pass in the context as a param here? @jhendrixMSFT this is performing an I/O operation, I think we should make it clear, what do you think?

File: README.md
Comment (jhendrixMSFT) [reply README.md:176]: Good catch, yes the directory name needs to change to match the package/module name.

File: go/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [reply go/azure_identity_credential_adapter.go:19]: You are correct, however we can't ship with this current implementation anyways (see my other comment about using the authentication policy).

File: README.md
Comment (jhendrixMSFT) [inline README.md:179]: github.com/jongio/azidext/go/azidext

File: go/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azure_identity_credential_adapter.go:1]: The convention is to place the source files in a directory with the same name as the package.

**Comment by jongio** [comment]
Thanks a lot @jhendrixMSFT and @catalinaperalta - Can you please approve this when ready to be merged?  I don't have much to add from a Go perspective, so will depend on you to give the final ok. Thanks,

File: go/azidext/go.mod
Comment (jhendrixMSFT) [inline go/azidext/go.mod:1]: the module identity should be fully qualified

`github.com/jongio/azidext/go/azidext`

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by jhendrixMSFT** [comment]
@Luyunmt thanks for all the iterations.

@catalinaperalta do you have any additional comments?

**Comment by jongio** [comment]
@catalinaperalta - Any other comments?  If not, could you please approve?  Thanks! Jon

File: go/azidext/azure_identity_credential_adapter_test.go
Comment (catalinaperalta) [inline go/azidext/azure_identity_credential_adapter_test.go:19]: Nit: the second "test" is redundant in the test name.

**Review by catalinaperalta — APPROVED** [review]
LGTM!

**Comment by jhendrixMSFT** [comment]
@jongio once this is merged can you please create the following tag for the commit.

`go/azidext/v0.1.0`

This will allow Go's module tooling to consume it.

File: go/azidext/azure_identity_credential_adapter.go
Comment (jhendrixMSFT) [inline go/azidext/azure_identity_credential_adapter.go:33]: We recently refactored some public surface area in `azcore` which precludes the need to pass a context here.  Can you please update `azcore` to `v0.10.0` and fix this accordingly?

**Review by jongio — APPROVED** [review]

**Comment by jhendrixMSFT** [comment]
Fixes https://github.com/Azure/azure-sdk-for-go/issues/11870

**Pull request merged**
